### ISSUE TITLE
Raspy Brick Orangutan - Tokens without UniV3 pairs or having low liquidity can be stolen by an attacker

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2023-06-tokemak-judging#841]
Raspy Brick Orangutan

high

# Tokens without UniV3 pairs or having low liquidity can be stolen by an attacker

## Summary

The `UniV3Swap` contract facilitates swaps on the UniV3, potentially across multiple pools. However, it does not account for situations where a pool might not exist or is illiquid. This enable attackers to create or manipulate pools, thereby determining the price, which can potentially drain the contract's funds.

## Vulnerability Detail

The `swap()` function allows swaps, It does this by encoding a `path` which is a sequence of token addresses and pool fees that represent the pools to be used in the swaps. There are several issues with this approach:

1. **Non-existent Pools:** There's no check to determine if a pool on the swap path exists. This allows an attacker to front-run the swap with the creation of a pool, setting any desired fee and price. After doing so, they can execute the swap at their preferred price, potentially sweeping the contract's funds. Importantly, an attacker can do this with any of the swaps after the path has been decoded.

2. **Low Liquidity Pools:** The contract does not check the liquidity of the pool. A pool with low liquidity can be easily manipulated in terms of price. Quoting prices from a low liquidity pool is not feasible and can lead to unfavorable swap rates. Multiple paths should ideally be considered to ensure the best quote is used.

## Impact

An attacker can manipulate pool prices or create new pools to exploit the contract, potentially resulting in substantial losses for the contract or unfavorable swap rates for users.

## Code Snippet

Relevant portion from can be seen [here](https://github.com/sherlock-audit/2023-06-tokemak/blob/5d8e902ce33981a6506b1b5fb979a084602c6c9a/v2-core-audit-2023-07-14/src/swapper/adapters/UniV3Swap.sol#L45-L65)

<details>
  <summary>Click to see code reference</summary>

```solidity
function swap(
    address routerAddress,
    address sellTokenAddress,
    uint256 sellAmount,
    address,
    uint256 minBuyAmount,
    bytes memory data
) external override onlyRouter returns (uint256) {

    IERC20(sellTokenAddress).safeApprove(routerAddress, sellAmount);

    IUniswapV3SwapRouter.ExactInputParams memory params = IUniswapV3SwapRouter.ExactInputParams({
        path: data,
        recipient: address(this),
        deadline: block.timestamp,
        amountIn: sellAmount,
        amountOutMinimum: minBuyAmount
    });

    return IUniswapV3SwapRouter(routerAddress).exactInput(params);
}
```

</details>

## Tool used

Manual Audit

## Recommendation

Before executing a swap, verify the existence of all pools on the path to protect against non-existent pool exploits.

Additionally it's advisable implement a mechanism to check the liquidity of a pool. Ensure the pool's liquidity meets a predefined threshold before performing swaps.

**Comment by Bauchibred** [comment]
Escalate

Do note that reason for escalation is that this issue has been misduped.

This issue seems to have been mistaken for the "missing deadline while executing swaps" based on the comment left by judge:

> non-issue/low, there are safe checks for the returned amount of LMPVaultRouter functions and BaseAsyncSwapper. Deadline doesn't cause losing funds


I argue that this is a different issue with likes of this been considered valid mediums in the past, one of this case is [this](https://github.com/sherlock-audit/2023-04-splits-judging/issues/26)

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> Do note that reason for escalation is that this issue has been misduped.
> 
> This issue seems to have been mistaken for the "missing deadline while executing swaps" based on the comment left by judge:
> 
> > non-issue/low, there are safe checks for the returned amount of LMPVaultRouter functions and BaseAsyncSwapper. Deadline doesn't cause losing funds
> 
> 
> I argue that this is a different issue with likes of this been considered valid mediums in the past, one of this case is [this](https://github.com/sherlock-audit/2023-04-splits-judging/issues/26)
> 

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by Trumpero** [comment]
I agree that this issue was misduped with an incorrect comment, that was my mistake. However, this issue should be categorized as low/info because it doesn't cause any loss of funds. The `UniV3Swap.swap()` function is only invoked via a delegate call from the SwapRouter contract, which already includes a safe check for the return amount of the swap.
https://github.com/sherlock-audit/2023-06-tokemak/blob/5d8e902ce33981a6506b1b5fb979a084602c6c9a/v2-core-audit-2023-07-14/src/swapper/SwapRouter.sol#L93-L124 
Additionally, the swap path for each swapper is set by the admin in the SwapRouter contract with `setSwapRoute` function, so any use of incorrect or malicious pairs for UniV3Swap should be considered the admin's mistake.

**Comment by hrishibhat** [comment]
@Bauchibred

**Comment by Bauchibred** [comment]
> Additionally, the swap path for each swapper is set by the admin in the SwapRouter contract with setSwapRoute function, so any use of incorrect or malicious pairs for UniV3Swap should be considered the admin's mistake.

**Comment by Bauchibred** [comment]
Appreciate  Judge's additional insight, only escalated cause issue was originally misduped, but now fully see how this would be considered admin mistake due to `setSwapRoute()`.

**Comment by Evert0x** [comment]
Planning to reject escalation and keep issue closed.

**Comment by Bauchibred** [comment]
Hi @Evert0x, based on the comment from judge 
> I agree that this issue was misduped with an incorrect comment, that was my mistake.

I believe the escalation  is valid, though since the issue is invalid, I'd argue that the right step to take is accept escalation and keep issue closed, no?

**Comment by Evert0x** [comment]
@Bauchibred I see that the comment left by the lead judge during the judging contest was very confusing. It makes sense that you created an escalation to seek clarification. 

However, since the issue is invalid. The escalation didn't provide any value to get us to the final judging state. 

I will reject the escalation because of this. 

Note: with duplication related issues that are edge cases to this.

**Comment by Evert0x** [comment]
result:
invalid
unique

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [Bauchibred](https://github.com/sherlock-audit/2023-06-tokemak-judging/issues/841/#issuecomment-1748798272): rejected
